data-prepper-plugins/aggregate-processor/src/test/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionSynchronizerTest.java

```diff
@@ -0,0 +1,153 @@
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import java.util.Optional;
+import java.util.concurrent.locks.Lock;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.mockito.Mockito.lenient;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class AggregateActionSynchronizerTest {
+
+    @Mock
+    private AggregateAction aggregateAction;
+
+    @Mock
+    private AggregateGroupManager aggregateGroupManager;
+
+    @Mock
+    private AggregateGroup aggregateGroup;
+
+    @Mock
+    private AggregateIdentificationKeysHasher.IdentificationHash identificationHash;
+
+    @Mock
+    private AggregateActionResponse aggregateActionResponse;
+
+    @Mock
+    private Lock concludeGroupLock;
+
+    @Mock
+    private Lock handleEventForGroupLock;
+
+    @Mock
+    private Event event;
+
+    @BeforeEach
+    void setup() {
+        lenient().doNothing().when(handleEventForGroupLock).lock();
+        lenient().doNothing().when(handleEventForGroupLock).unlock();
+        lenient().doNothing().when(concludeGroupLock).unlock();
+        lenient().doNothing().when(aggregateGroup).clearGroupState();
+        lenient().doNothing().when(aggregateGroup).resetGroupStart();
+        lenient().doNothing().when(aggregateGroupManager).putGroupWithHash(identificationHash, aggregateGroup);
+        lenient().doNothing().when(aggregateGroupManager).removeGroupWithHash(identificationHash, aggregateGroup);
+        when(aggregateGroup.getConcludeGroupLock()).thenReturn(concludeGroupLock);
+        when(aggregateGroup.getHandleEventForGroupLock()).thenReturn(handleEventForGroupLock);
+    }
+
+    private AggregateActionSynchronizer createObjectUnderTest() {
+        final AggregateActionSynchronizer.AggregateActionSynchronizerProvider aggregateActionSynchronizerProvider = new AggregateActionSynchronizer.AggregateActionSynchronizerProvider();
+        return aggregateActionSynchronizerProvider.provide(aggregateAction, aggregateGroupManager);
+    }
+
+    @Test
+    void concludeGroup_with_tryLock_false_returns_empty_optional() {
+        final AggregateActionSynchronizer objectUnderTest = createObjectUnderTest();
+        when(concludeGroupLock.tryLock()).thenReturn(false);
+
+        final Optional<Event> concludeGroupEvent = objectUnderTest.concludeGroup(identificationHash, aggregateGroup);
+
+        verifyNoInteractions(handleEventForGroupLock);
+        verifyNoInteractions(aggregateAction);
+        verifyNoInteractions(aggregateGroupManager);
+        verify(aggregateGroup, times(0)).resetGroupStart();
```

Comment:
These verifications are not performed in order.

But, Mockito does support in-order verification. Here is [an example](https://github.com/opensearch-project/data-prepper/blob/1183a7a13517885021c023971c77514f654dcc95/data-prepper-plugins/peer-forwarder/src/test/java/com/amazon/dataprepper/plugins/prepper/peerforwarder/discovery/AwsCloudMapPeerListProviderTest.java#L337-L345) from Data Prepper. This example uses the BDD-style then-should functions. You can still use verify with something like: `inOrder.verify(aggregateGroup, times(0)).clearGroupState();`

This comment applies to a number of verifications in this file.

Reply:
That's cool. I learned something new.